dftimewolf/lib/collectors/gsheets.py

```diff
@@ -0,0 +1,243 @@
+  """Collector for entries from Google Sheets. """
+
+  SCOPES = ['https://www.googleapis.com/auth/spreadsheets.readonly']
+
+  _CREDENTIALS_FILENAME = '.dftimewolf_google_sheets_credentials.json'
+  _CLIENT_SECRET_FILENAME = '.dftimewolf_google_sheets_client_secret.json'
+
+  def __init__(self,
+               state: 'state.DFTimewolfState',
+               name: Optional[str] = None,
+               critical: bool = False) -> None:
+    """Initializes a Google Sheets collector."""
+    super(GoogleSheetsCollector, self).__init__(
+        state, name=name, critical=critical)
+    self._credentials = None
+    self._spreadsheet_id = ''
+    self._sheets_names = ''
+    self._mandatory_columns = ['message', 'datetime', 'timestamp_desc']
+    self._all_sheets = False
+
+  # pylint: disable=arguments-differ
+  def SetUp(self, spreadsheet: str, sheets_names: List[str]) -> None:
+    """Sets up a a Google Sheets collector.
+
+    Args:
+      spreadsheet: ID or URL of the sheet to pull data from
+      sheets_names: List of sheets names inside the spreadsheet to parse. 'All'
+        will parse all sheets inside a spreadsheet.
+    """
+    self._credentials = self._GetCredentials()
+    self._spreadsheet_id = self._ValidateSpreadSheetId(spreadsheet)
+    self._sheets_names = sheets_names
+    if 'All' in sheets_names:
+      self._all_sheets = True
+
+  def Process(self) -> None:
+    """Copies entries from Google Sheets."""
+
+    try:
+      # Retrive list of sheets in the spreadsheet
+      service = self._BuildSheetsResource(self._credentials)
+      result = service.spreadsheets().get(
+          spreadsheetId=self._spreadsheet_id).execute()
+      spreadsheet_title = result.get('properties', {}).get('title')
+      sheets = (result.get('sheets', []))
+
+      for sheet in sheets:
+        if not sheet.get('properties'):
+          continue
+
+        sheet_title = sheet.get('properties', {}).get('title')
+
+        if not self._all_sheets and sheet_title not in self._sheets_names:
+          continue
+
+        self.logger.info('Parsing sheet: {0:s}'.format(sheet_title))
+
+        df = self._ParseSheet(self._spreadsheet_id, sheet_title)
+
+        if df is None or df.empty:
+          continue
+
+        output_file = tempfile.NamedTemporaryFile(
+            mode='w', delete=False, encoding='utf-8', suffix='.csv')
+        output_path = output_file.name
+        self.logger.info(
+            'Downloading results to of sheet "{0:s}" to {1:s}'.format(
+                sheet_title, output_path))
+
+        df.to_csv(index=False, na_rep='NaN', path_or_buf=output_file)
+
+        self.logger.success(
+            'Downloaded results of sheet "{0:s}" to {1:s}'.format(
+                sheet_title, output_path))
+        output_file.close()
+        sheet_csv_file = containers.File(
+            name='{0:s}_{1:s}'.format(spreadsheet_title, sheet_title),
+            path=output_path)
+        self.state.StoreContainer(sheet_csv_file)
```

Comment:
Do we want to store CSV files, or pandas dataframes? I'd think the latter might be better (TimesketchExporter also supports pandas dataframes IIRC).

Reply:
The timesketch exporter calls "add_file" from the "ImportStreamer" class. This method expects a csv or json file. The "ImportStreamer" accepts DataFrames ("add_data_frame" method), but the exporter module doesn't support that.

Reply:
The timesketch exporter calls "add_file" from the "ImportStreamer" class. This method expects a csv or json file. The "ImportStreamer" accepts DataFrames ("add_data_frame" method), but the exporter module doesn't support that.